File: Swap-Nodes-in-Pairs/Swap-Nodes-in-Pairs.cpp

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        if(!head || !head->next)
            return head;
        ListNode *s=head;
        ListNode *p=head;
        ListNode *prev=NULL;
        ListNode *phead=NULL;
        while(p)
        {
            if(!p->next)
                return phead;
            else
            {
                s=p;
                p=p->next;
                ListNode *t=p->next;
                if(prev)
                    prev->next=p;
                p->next=s;
                s->next=t;
                if(!phead) 
                    phead=p;
                prev=s;
                p=t;
            }
        }
        return phead;
    }
};


 

File: Swap-Nodes-in-Pairs/Swap-Nodes-in-Pairs.py

‘’’
 Definition for singly-linked list.
 class ListNode:
     def __init__(self, x):
         self.val = x
         self.next = None
‘’’
class Solution:
    # @param a ListNode
    # @return a ListNode
    def swapPairs(self, head):
        if head==None:
            return None
        if head.next==None:
            return head
        p=head
        q=head.next
        p.next=q.next
        q.next=p
        head=q
        while p!=None:
            s=p
            p=p.next
            if p==None:
                break
            q=p.next
            if q==None:
                s.next=p
                break
            p.next=q.next
            q.next=p
            s.next=q
        return head
 